libcontainer/cgroups/devices/systemd.go

```
@@ -126,7 +140,11 @@ func systemdProperties(r *configs.Resources, sdVer int) ([]systemdDbus.Property,
 				}
 				if group == "" {
 					// Couldn't find a group.
-					logrus.Warnf("could not find device group for '%v/%d' in /proc/devices -- temporarily ignoring rule: %v", rule.Type, rule.Major, *rule)
+					err = fmt.Errorf("systemd older than v240 does not support wildcard-minor rules for devices not listed in /proc/devices: +%v", *rule)
+					if cgroupVer == 2 {
+						return nil, err
+					}
```

Surely we don't want to always return an error in this case, even if systemd supports it? I guess it's more consistent but now we're going to start returning errors...

---

This here is a combination of these four factors:
1. The rule is minor-wildcard (i.e. $MAJOR:*).
2. Systemd is older than v240 (so we can't use {block,char}-NNN syntax, where NNN == $MAJOR).
3. The MAJOR specified can't be found in /proc/devices (so we can't use {block,char}-NAME syntax, where NAME is an entry from /proc/devices corresponding to specified major number).
4. Cgroup is v2, meaning we solely rely on systemd for rules management (see the change to libcontainer/cgroups/fs2/fs2.go), meaning we can't enable access to these devices.

So, the two options we have are ignore this device rule (with a warning) or return an error. I'd rather do the latter since otherwise it's harder to debug cases when we can't access a device.

NOTE we do not return errors if we think systemd supports it.

---

@cyphar It just occurred to me you meant we should still generate our own EBPF (in this case these errors might be downgraded to warnings).

Or do something else -- say, if there are no warnings when generating systemd device rules, skip adding our own EBPF, otherwise add one?

---

I played with this a bit, checking a case when a rule is not added to systemd DeviceAccess, but only to the EBPF we generate. In this case kernel denies access to the device.

This essentially means:
 - if we can't translate a device rule to systemd lingo, having our own EBPF won't help to allow access;
 - if we can translate all the rules, and systemd is bug free™, having our own EBPF is redundant.
 
 This is the reason why in this PR I chose to
 - return an error if a rule can't be applied;
 - not generate a second EBPF.
 
 Now, we can certainly downgrade the errors to warnings, let me know if this is how you prefer it.

---

@cyphar OK I've downgraded errors to warnings. PTAL.

---

Yeah, I would've preferred to output a warning for this case because right now we silently ignore such cases and create BPF filters that don't do anything, but changing it to an error would be too extreme for such users imho.